### PR TITLE
User Fileservers

### PR DESCRIPTION
**Pull request by athornton**
This merges in the actual content of the technote.

### CODE REVIEW

File: index.rst [hunk 1]
```diff
@@ -13,17 +12,211 @@
...
+
+The user fileserver adds three routes to the controller.
+
+``GET /files``
+    Creates fileserver objects for user if necessary and returns usage
+    text.  That text instructs a user how to acquire a token for the
+    fileserver and tells the user to direct a WebDAV client with that
+    token to ``/files/<username>``.
+
+  Credential scopes required: ``exec:notebook``
```
Comment (rra) [inline index.rst]: I think the inconsistent indentation isn't going to render well.
```suggestion
    Credential scopes required: ``exec:notebook``
```

File: index.rst [hunk 2]
```diff
@@ -13,17 +12,211 @@
...
+
+``GET /nublado/fileserver/v1/users``
+    Returns a list of all users with running fileservers.
+    Example:
+
+    .. code-block:: json
+
+       [ "adam", "rra" ]
+
+  Credential scopes required: ``admin:jupyterlab``
```
Comment (rra) [inline index.rst]:
```suggestion
    Credential scopes required: ``admin:jupyterlab``
```

File: index.rst [hunk 3]
```diff
@@ -13,17 +12,211 @@
...
+
+``DELETE /nublado/fileserver/v1/<username>``
+    Removes fileserver objects (if any) for the specified user.
+
+  Credential scopes required: ``admin:jupyterlab``
```
Comment (rra) [inline index.rst]:
```suggestion
    Credential scopes required: ``admin:jupyterlab``
```

**Review by rra — APPROVED** [review]
A few minor formatting nits. This lays out the basics well, although I'll channel Frossie and say that it would be nice to have a diagram or two. Specifically, a block diagram of how the user's request flows through to the lab controller on the first request alongside showing how it flows to the fileserver pod once that's been spawned would help show how there's a catch-all ingress but then the requests are handed over to a more specific ingress.

It might also be worth mentioning the period where the fileserver shuts down due to inactivity but still has a valid ingress and why that's not a concern (because we use a watch and delete the ingress almost immediately).

Also, I think you said you mounted all the file systems from the notebook configuration into the fileserver pod but under a shared directory so that the WebDAV server would see them all. That may be worth describing here in a bit more detial so that it's clear how we retain the same user-visible paths in both the file server and the lab.

**Comment by athornton** [comment]
Diagrams added.  Discussion of watch added.  (Minimal) discussion of mountpoint aggregation added.

**Pull request merged**